Containerd cadvisor integration

abhi opened the pull request:
This commit includes : 
1) changes to integrate containerd runtime to cadvisor to collect container stats.
2) test cases.

cc @Random-Liu

abhi commented:
I signed it!

Random-Liu commented:
/ok-to-test

Random-Liu commented:
@dashpole Could you help take a look when you have time? Thanks!

Random-Liu commented:
/test pull-cadvisor-e2e

abhi commented:
/retest

Random-Liu commented on Godeps/Godeps.json:
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "github.com/google/cadvisor",
-	"GoVersion": "go1.8",
+	"GoVersion": "go1.9",
```
Can you use go1.8?

Random-Liu commented on Godeps/Godeps.json:
```diff
@@ -610,36 +687,40 @@
-		{
 			"ImportPath": "golang.org/x/net/context",
-			"Rev": "5627bad10b821d907ac4dc81fb3535ec7dd81c2b"
+			"Rev": "7dcfb8076726a3fdd9353b6b8a1f1b6be6811bd6"
```
Why do we need to update `golang.org/x/net`?

Random-Liu commented:
@abhi 
```
W1108 17:12:00.427] vendor/github.com/sirupsen/logrus/text_formatter.go:13:2: cannot find package "golang.org/x/crypto/ssh/terminal" in any of:
W1108 17:12:00.428] 	/go/src/github.com/google/cadvisor/vendor/golang.org/x/crypto/ssh/terminal (vendor tree)
W1108 17:12:00.428] 	/usr/local/go/src/golang.org/x/crypto/ssh/terminal (from $GOROOT)
W1108 17:12:00.428] 	/go/src/golang.org/x/crypto/ssh/terminal (from $GOPATH)
```

Not all dependencies are included.

Random-Liu commented:
/test pull-cadvisor-e2e

abhi commented:
/test pull-cadvisor-e2e

Random-Liu commented:
@abhi The godep change is merged. Please rebase your PR.

The test is passing. I think ready for a review now. :)

abhi commented:
@Random-Liu Done. ready for review.

Random-Liu commented on container/containerd/client_test.go:
```diff
@@ -0,0 +1,53 @@
+	}
+	cntr, ok := c.cntrs[id]
+	if !ok {
+		return nil, fmt.Errorf("unable to find container %s", id)
```
nit: s/%s/%q

Random-Liu commented on container/containerd/factory.go:
```diff
@@ -0,0 +1,151 @@
+// The namespace under which containerd aliases are unique.
+const (
+	ContainerdNamespace = "containerd"
+	RootFSStorageDir    = "/var/run/containerd/io.containerd.runtime.v1.linux/k8s.io/"
```
nit: Use `k8sContainerdNamespace`?

Random-Liu commented on container/containerd/grpc.go:
```diff
@@ -0,0 +1,50 @@
+	if !ok {
+		ctx = namespaces.WithNamespace(ctx, ni.namespace)
+	}
+
```
nit: unnecessary empty line.

Random-Liu commented on container/containerd/factory.go:
```diff
@@ -0,0 +1,151 @@
+	ctx := context.Background()
+	_, err := self.client.LoadContainer(ctx, id)
+	if err != nil {
+		return false, true, fmt.Errorf("failed to load container: %v", err)
```
Why `canHandle=false` and `canAccept=true`? Don't quite understand. :p 
And should we load task instead, and check whether task is running?

Random-Liu commented on container/containerd/factory_test.go:
```diff
@@ -0,0 +1,75 @@
+
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/typeurl"
+	containerlibcontainer "github.com/google/cadvisor/container/libcontainer"
```
nit: Separate group for `cadvisor`.

Random-Liu commented on integration/runner/runner.go:
```diff
@@ -44,7 +44,7 @@ const (
 )
 
-var cadvisorTimeout = flag.Duration("cadvisor_timeout", 15*time.Second, "Time to wait for cAdvisor to come up on the remote host")
+var cadvisorTimeout = flag.Duration("cadvisor_timeout", 30*time.Second, "Time to wait for cAdvisor to come up on the remote host")
```
I still a bit concern. Have you tried without containerd running? How long does it take to try connecting containerd?

abhi replied:
I am going to get back on this. Running on a single machine doesnt seem to be an issue . Its only when the CI runs ssh and runs it that the timeout needs to be increased. Will come back with a valid explaination

Random-Liu commented on fs/fs.go:
```diff
@@ -90,9 +90,10 @@ type RealFsInfo struct {
+	Docker     DockerContext
+	RktPath    string
+	Crio       CrioContext
+	Containerd ContainerdContext
```
Do we need this change? It seems that we are not using it.

Random-Liu commented on container/containerd/handler.go:
```diff
@@ -0,0 +1,244 @@
+	}
+	rootfs := "/"
+	if !inHostNamespace {
+		rootfs = RootFSStorageDir + cntr.ID + spec.Root.Path
```
This doesn't seem right to me. What rootfs do we need here? It seems that we are using this rootfs to get network stats.
However,
1) Why docker/cri-o are using `/rootfs`?
2) There may not be `proc` inside container, right? e.g. an image built from scratch.

abhi replied:
if its in non host namespace then proc is located at ```/var/run/containerd/io.containerd.runtime.v1.linux/k8s.io/<containerid>/rootfs```. This from what I verified during coding. Am I wrong here ?

Random-Liu replied:
TBH, I don't know.
The pid we get is host pid, right? Why don't we just access `/proc/pid/net/ns`?

Random-Liu replied:
I believe the `inHostNameSpace` here, means whether cadvisor is running in host namespace or inside container.

When cadvisor is running inside a container, we mount host root filesystem to `/rootfs`.

Random-Liu replied:
So we should have same code with docker, and remove the container rootfs related logic.

Random-Liu replied:
See [here](https://github.com/google/cadvisor/blob/master/manager/manager.go#L191-L196).

Random-Liu commented on container/containerd/handler.go:
```diff
@@ -0,0 +1,244 @@
+func (self *containerdContainerHandler) Cleanup() {
+}
+
+func (self *containerdContainerHandler) GetContainerIPAddress() string {
```
This is for custom metrics. We need to figure out whether it's ok to not support it.
@dashpole Are we still using custom metrics?

dashpole replied:
I believe we still support it...  IIRC from my last conversation with the warsaw folks, we probably wont support this model forever, but I believe we still support it ATM

Random-Liu replied:
@dashpole I don't think we could get ip of docker container in cadvisor, when we are using kubenet. So do we really support this in GKE?

Random-Liu replied:
Thanks @dashpole @DirectXMan12!
Confirmed that this doesn't work anymore. We don't need to support it.

Random-Liu commented on container/containerd/handler.go:
```diff
@@ -0,0 +1,244 @@
+}
+
+func (self *containerdContainerHandler) GetSpec() (info.ContainerSpec, error) {
+	hasFilesystem := !self.ignoreMetrics.Has(container.DiskUsageMetrics)
```
What does this mean? I saw both cri-o and docker have this.
However, because we don't collect image filesystem stats, so just make sure.

dashpole replied:
do you collect container disk stats?  If not, you should just set this to false.  I believe this is meant to indicate whether or not this container has disk stats.

dashpole commented:
FYI, I plan to cut cadvisor v0.28.1, which will correspond to k8s v1.9.0 this Friday, 11/17.

abhi commented:
@dashpole thanks for the heads up.

abhi commented on container/containerd/client.go:
```diff
@@ -0,0 +1,123 @@
+	})
+	if err != nil {
+		err = errdefs.FromGRPC(err)
+		if errdefs.IsNotFound(err) {
```
Done

Random-Liu replied:
Still. Don't understand this one.

Random-Liu commented on container/containerd/factory.go:
```diff
@@ -0,0 +1,150 @@
+var ArgContainerdEndpoint = flag.String("containerd", "unix:///var/run/containerd.sock", "containerd endpoint")
+
+// The namespace under which containerd aliases are unique.
+const (
```
Remove brackets.

Random-Liu commented on container/containerd/factory.go:
```diff
@@ -0,0 +1,150 @@
+	// If container and task lookup in containerd fails then we assume
+	// that the container state is not known to containerd
+	ctx := context.Background()
+	_, err := self.client.LoadContainer(ctx, id)
```
Check running?

Random-Liu commented on fs/fs.go:
```diff
@@ -105,6 +105,10 @@ type CrioContext struct {
 	Root string
 }
 
+type ContainerdContext struct {
```
Never used?

Random-Liu commented:
@abhi LGTM other than `hasFilesystem`. Because I'm not quite sure what that means. :)

dashpole commented:
It doesnt need to happen now, but eventually it would be nice to have disk stats for containerd through cAdvisor.  k8s would disable disk stats, but it would mean that standalone cAdvisor could still be used to monitor docker after docker rebases on top of containerd.

This PR lgtm other than the one nit.

abhi commented:
@dashpole @Random-Liu thanks for reviewing. I have addressed the comments and rebased. 

@dashpole : also current docker master already uses containerd 1.0 beta. But we would still use the docker apis to collect stats for standalone cadvisor case ?

Random-Liu commented on container/containerd/client.go:
```diff
@@ -0,0 +1,122 @@
+		grpc.WithDialer(dialer.Dialer),
+		grpc.WithBlock(),
+		grpc.WithTimeout(2 * time.Second),
+		grpc.WithBackoffMaxDelay(3 * time.Second),
```
Why do you add this back?

abhi replied:
to avoid the retries by the dialer. So I set the timeout and use withblock and backoffdelay.

Random-Liu commented:
> It doesnt need to happen now, but eventually it would be nice to have disk stats for containerd through cAdvisor. k8s would disable disk stats, but it would mean that standalone cAdvisor could still be used to monitor docker after docker rebases on top of containerd.

@dashpole When docker move to containerd 1.0 image management, I guess the image fs stats in the current docker integration won't work. We need to either change cadvisor to understand the new disk layout, or let cadvisor get fs stats from underlying containerd.

@abhi I see that we hard code containerd unix socket path now. In the future, after docker moves to the new containerd, if it is using the same socket path, current cadvisor code might get confused.
The current containerd integration is mainly for `k8s.io` containerd namespace, so the containerd namespace may help us avoid trouble. However, we should keep an eye on that.

abhi commented:
@Random-Liu yes thats right. We are currently covered by the namespace seperation. We can revisit for docker+containerd support
